File: 算法与数据结构/python实现数据结构/linkedlist_/linkedqueue.py
# coding:utf-8


class Node(object):
    def __init__(self, e, next_=None):
        self.e = e
        self.next = next_


class LinkedQueue(object):
    """
    基于双向链表的队列
    """
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def is_empty(self):
        """
        队列是否为空
        :return:
        """
        return self.size == 0

    def get_size(self):
        """
        队列中元素的个数
        :return:
        """
        return self.size

    def enqueue(self, e):
        """
        元素入队, 从双向链表末尾添加
        :return:
        """
        if self.tail is None:
            self.tail = Node(e)
            self.head = self.tail
        else:
            new_node = Node(e)
            self.tail.next = new_node
            self.tail = self.tail.next
        self.size += 1

    def dequeue(self):
        """
        元素出队，从双向链表表头出队
        :return:
        """
        if self.is_empty():
            raise ValueError("dequeue failed, stack is empty")
        ret_node = self.head
        self.head = ret_node.next
        ret_node.next = None
        if self.head is None:
            self.tail = None
        self.size -= 1

        return ret_node.e

    def get_front(self):
        """
        查询队列队顶元素信息
        :return:
        """
        if self.is_empty():
            raise ValueError("queue is empty")
        return self.head.e

    def __repr__(self):
        res = "Queue: front "
        cur = self.head
        while cur:
            res += "{0}->".format(cur.e)
            cur = cur.next
        res += "None tail"

        return res


def test_linked_queue():
    q = LinkedQueue()
    for i in range(10):
        q.enqueue(i)
        print(q)
    q.dequeue()
    print(q)
    q.dequeue()
    print(q)
    q.dequeue()
    print(q)


if __name__ == "__main__":
    test_linked_queue()

File: 算法与数据结构/golang实现数据结构/DataStructAndAlgorithm/linkedlist/linkedqueue/linkedqueue.go
package linkedqueue

import (
	"errors"
	"fmt"
)

type Node struct {
	e interface{}
	next *Node
}

// linklist有dummyHead，但是没有尾结点，导致出队需要遍历，导致O(n)，这里重新定义，添加尾结点tail，而不复用
type LinkedQueue struct {
	head *Node
	tail *Node
	size int
}

// 新建节点对象
func NewNode(e interface{}, next *Node) *Node {
	return &Node{e: e, next: next}
}

// 获取队列元素个数
func (l *LinkedQueue) GetSize() int {
	return l.size
}

// 判断队列是否为空
func (l *LinkedQueue) IsEmpty() bool {
	return l.size == 0
}

// 新建链表队列对象
func NewLinkedQueue() * LinkedQueue {
	queue := LinkedQueue{}
	return &queue
}

// 元素入队，从链表尾部进行
func (l *LinkedQueue) Enqueue(e interface{}) (err error) {
	if l.tail == nil {
		l.tail = NewNode(e, nil)
		l.head = l.tail
	} else{
		newNode := NewNode(e, nil)
		l.tail.next = newNode
		l.tail = newNode
	}
	l.size ++
	return err
}

// 元素出队, 从链表头部出队
func (l *LinkedQueue) Dequeue() (e interface{}, err error) {
	if l.IsEmpty() {
		return nil, errors.New("cannot dequeue from an empty queue")
	}
	retNode := l.head
	l.head = l.head.next
	retNode.next = nil
	if l.head == nil{
		l.tail = nil
	}
	l.size --
	return  retNode.e, err
}

// 查询队首元素信息
func (l *LinkedQueue) GetFront() (e interface{}, err error){
	if l.IsEmpty() {
		err = errors.New("queue is empty")
		return nil, err
	}
	return l.head.e, err
}

// 格式化输出队列信息
func (l *LinkedQueue) String() string {
	res := "Queue: front "
	cur := l.head
	for cur != nil {
		res += fmt.Sprintf("%v->", cur.e)
		cur = cur.next
	}
	res += "nil tail"
	return res
}
